### DISCUSSION TITLE
会話文の中に変数を表示させるにはどうしたら良いでしょうか？

### DISCUSSION THREAD

**Discussion by mikuro898** [HANON-games/Shiden#19] [Q&A] [answered]
いつもお世話になっています。
Wikiを見ても初心者の私では理解できなかったので質問させていただきます。
Wiki>コマンド一覧>Textコマンド   に「テキストには { 変数名 } で変数の値を入れられます。」と記載されているのですが、
この変数名とは具体的にどの型の変数でしょうか？またどのブループリントクラス(BP_ShidenTextCommandやAC_MyShidenActorなど)にその変数を設定させればよいでしょうか？
見当違いな質問でしたら申し訳ありません。

具体的にやりたいことは主人公の名前をプレイヤーが決めて、その名前を登場人物が使うといった様にしたいです。主人公の名前を変数として定義する機能はすでに作ってあるので、その変数をテキスト上に自然な形で出せないかなと考えています。

**Comment by HANON-games** [comment]
@mikuro898 
Shiden を利用いただきありがとうございます。

Shiden における変数とは AssignVariable コマンド等を使用することでシナリオ内で定義するものになります。

ブループリントで定義した変数を使用するには、文字列の場合、以下のように AddUserString 関数 (ユーザセーブに反映したい場合) や AddSystemString 関数 (システムセーブに反映したい場合) を利用いただくことになるかと思います。
セーブデータのロード時に Shiden で管理している変数がリセットされるため、セーブロード機能がある場合は必ずロード後に追加してください (コマンドを自作してその中で変数を設定するのが一番やりやすいかもしれません)。

![image](https://github.com/user-attachments/assets/71c99364-e9b3-4cce-bde0-425fa48d1d75)

上記で設定した変数を使用したい場合は以下のようにテキストを入力します (今月公開予定の新しいベータバージョンのスクショのため、現在公開しているものとは若干項目が異なりますがご了承ください... 🙏 )。

![image](https://github.com/user-attachments/assets/1c23479c-d605-4834-8ec1-87ee52387af2)

なお、次のベータバージョンにてプレイヤーの入力を受け付ける TextInput コマンドを追加する予定です。
詳細につきましては下記をご覧いただけますと幸いです。

https://github.com/HANON-games/Shiden/discussions/17

**Reply by mikuro898** [reply, marked as answer]
ありがとうございます！解決しました！！

一応忘れないうちに書き留めておきます。
![スクリーンショット 2024-11-08 025656](https://github.com/user-attachments/assets/7d12e1fd-9466-45ec-a8c8-ce41328bfd9e)
↑自作コマンド
![image](https://github.com/user-attachments/assets/6f2b4039-e97f-43f4-89ab-282cc05795b2)
↑AC_MyShidenActor
![image](https://github.com/user-attachments/assets/4975a13a-47c0-40f8-a43a-b28782fced67)
↑Game Instance
![image](https://github.com/user-attachments/assets/b5a02bf2-8b82-4946-a30d-49b777014771)
↑UMG

色々試したのですが、私の実力ではグローバル変数やブループリントインターフェイスを使った方法では変数を直接コマンドに継承することが出来ませんでした。ただAC_MyShidenActorを経由すれば遠回りですが上手くいきました。今回UMG上に目的の変数があったため、UMG→Game Instance→AC_MyShidenActor→コマンドの順に変数を継承させて半ばゴリ押しでやっています。しかし正常に動いているので満足です。
